### PR TITLE
Add find_address tool

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds a `find_address` tool, which can be used to find 'invalid' public keys with a certain desired prefix.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Review by bkase — APPROVED** [review]
Cool — we should get a crypto review to double check the invalidity of the private key part

File: src/lib/signature_lib/find_address/find_address.ml [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+open Core_kernel
+open Signature_lib
+
+let debug = false
+
+let default_prefix = "zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz"
+
+let desired_prefix =
+  try Sys.argv.(1)
```
Comment (psteckler) [inline src/lib/signature_lib/find_address/find_address.ml:23]: should there be a check that the command line argument contains only letters from `Base58_check.mina_alphabet`?

Comment (mrmr1993) [reply src/lib/signature_lib/find_address/find_address.ml:23]: I wanted to add this, but there didn't seem to be an easy way to get back the string from a `Base58.alphabet`, so I left it without.

Comment (psteckler) [reply src/lib/signature_lib/find_address/find_address.ml:23]: A hack would be to run `B58.decode` on each character, given that alphabet, see if it raises.

**Comment by mrmr1993** [comment]
Updated. From a sample run:
```
$ dune exec src/lib/signature_lib/find_address/find_address.exe -- burnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz | head
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzUEGqXVGH4 (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzUEGzkeuED (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzAeRtPV2GX (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzAeS38qZ5p (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzmp7r7UN6X (valid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzmp7tmpty1 (valid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz1rWSdEfqR (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz1rWWdSHie (invalid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzd2CJJhGjn (valid)
B62qiburnzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzd2CTndVxq (valid)
```

File: src/lib/signature_lib/find_address/find_address.ml [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+    A Mina public key is a point on the Pallas elliptic curve, which we can
+    represent as a pair of Pallas base field elements `(x, y)`. These values
+    must satisfy the elliptic curve equation `y^2 = x^3 + 5` over the base
+    field.
+
+    In order to reduce the amount of data that is sent over the Mina network,
+    and to reduce the length of base58-check encoded public keys, the Mina key
+    format uses a 'compressed' representation.
+    This compressed representation is the coorindate `x` of the public key and
```
Comment (mimoo) [inline src/lib/signature_lib/find_address/find_address.ml:26]: "coorindate"

**Review by mimoo — APPROVED** [review]
LGTM!

File: src/lib/signature_lib/find_address/find_address.ml [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    an `is_odd` boolean, where `is_odd` determines which value of `y` it
+    matches, either `y = sqrt(x^3 + 5)` or `y = -sqrt(x^3 + 5)`.
+
+    For some values of `x`, there is no square-root of `x^3 + 5` in the finite
+    field, and so there is no Mina public key with that `x` coordinate.
+    However, since we can choose any value of `x` in the 'compressed'
+    representation, it is possible to encode one of these values.
+    We call a compressed public key with no `y` coordinate 'invalid', and a
+    compressed public key where we can calculate the `y` coordinate 'valid'.
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/signature_lib/find_address/find_address.ml:35]: Indeed, we fixed ours not long before the mainnet launch.

File: src/lib/signature_lib/find_address/find_address.ml [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    See https://en.wikipedia.org/wiki/Elliptic_curve_point_multiplication for
+    more information.
+
+    The reverse direction -- transforming a public key into a private key -- is
+    considered to be computationally intractable. This problem is generally
+    known as the 'discrete logarithm problem', and is the basis for the
+    security of Mina's signature scheme. The best known algorithms for finding
+    a private key from a public key are only slightly better than
+    'guess-and-check', where you repeatedly choose a random number and test it
+    to check whether it is right.
```
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/signature_lib/find_address/find_address.ml:30]: Discussed out-of-band; I've updated to mention the algorithm and give probabilities based on that instead.

**Review by bkase — APPROVED** [review]
Beautiful

**Comment by yourbuddyconner** [comment]
This pull request has been mentioned on **MinaResearch**. There might be relevant details there:

https://forums.minaprotocol.com/t/proposal-to-remove-all-tokens-from-burn-address-at-next-hardforks/5738/1

**Pull request merged**